token/program-2022/src/instruction.rs

```diff
@@ -409,6 +434,127 @@ pub enum TokenInstruction {
+    /// Initialize the transfer fee on a new mint.
+    ///
+    /// Fails if the mint has already been initialized, so must be called before
+    /// `InitializeMint`.
+    ///
+    /// The mint must have exactly enough space allocated for the base mint (82
+    /// bytes), plus 83 bytes of padding, 1 byte reserved for the account type,
+    /// then space required for the extension.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   0. `[writable]` The mint to initialize.
+    InitializeMintTransferFee {
+        /// Pubkey that may update the fees
+        fee_config_authority: COption<Pubkey>,
+        /// Harvest instructions must be signed by this key
+        fee_withdraw_authority: COption<Pubkey>,
+        /// Amount of transfer collected as fees, expressed as basis points of the
+        /// transfer amount
+        transfer_fee_basis_points: u16,
+        /// Maximum fee assessed on transfers
+        maximum_fee: u64,
+    },
+    /// Transfer, providing expected mint information and fees
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   * Single owner/delegate
+    ///   0. `[writable]` The source account. Must include the `AccountTransferFeeExtension`.
+    ///   1. `[]` The token mint. Must include the `AccountMintFeeExtension`.
+    ///   2. `[writable]` The destination account. Must include the `AccountTransferFeeExtension`.
+    ///   3. `[signer]` The source account's owner/delegate.
+    ///
+    ///   * Multisignature owner/delegate
+    ///   0. `[writable]` The source account.
+    ///   1. `[]` The token mint.
+    ///   2. `[writable]` The destination account.
+    ///   3. `[]` The source account's multisignature owner/delegate.
+    ///   4. ..4+M `[signer]` M signer accounts.
+    TransferCheckedWithFee {
+        /// The amount of tokens to transfer.
+        amount: u64,
+        /// Expected number of base 10 digits to the right of the decimal place.
+        decimals: u8,
+        /// Expected fee assessed on this transfer, calculated off-chain based on
+        /// the transfer_fee_basis_points and maximum_fee of the mint.
+        fee: u64,
+    },
+    /// Transfer all withheld tokens to an account. Signed by the mint's
+    /// fee withdraw authority.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   * Single owner/delegate
+    ///   0. `[writable]` The token mint. Must include the `MintTransferFeeExtension`.
+    ///   1. `[writable]` The fee receiver account. Must include the `AccountTransferFeeExtension`.
+    ///      associated with the provided mint.
+    ///   2. `[signer]` The mint's `fee_withdraw_authority`
+    ///
+    ///   * Multisignature owner/delegate
+    ///   0. `[writable]` The token mint.
+    ///   1. `[writable]` The destination account.
+    ///   2. `[]` The mint's `fee_withdraw_authority`'s multisignature owner/delegate.
+    ///   3. ..3+M `[signer]` M signer accounts.
+    HarvestWithheldTokensFromMint,
```

Comment:
I got confused here. Based on the doc and instruction name -- esp the word "Harvest" -- I thought this instruction was the same as `HarvestWithheldTokensToMint`, except it sends the tokens to an account instead of the mint. But since no source accounts are provided, I must be wrong.

Perhaps this should be `WithdrawWithheldTokensFromMint`?

Reply:
That makes a lot of sense, we use the term `WithdrawWithheld` everywhere else.  The idea is there's one instruction to harvest from accounts to the mint, and another to withdraw from the mint to an account.

For better parallelization, we should probably also add a `Withdraw` instruction to withdraw straight from a bunch of accounts and into another one, which wouldn't require a writable mint.  What do you think?

Reply:
Yeah I like that idea. The withdraw to mint operation is really just to facilitate emptying all the withheld tokens before closing an account.  If we can assert that an account with 0 tokens will always have 0 withheld tokens then we could remove withdrawing to mint entirely

Reply:
Done.

>  If we can assert that an account with 0 tokens will always have 0 withheld tokens then we could remove withdrawing to mint entirely

I can't come up with a model that guarantees this :thinking:

Reply:
Excellent, these changes allay all my confusion :)

> I can't come up with a model that guarantees this 🤔

I was going to suggest moving the sender's withheld_amount to the destination account, but I think I just went through the same thought process you two did [here](https://github.com/solana-labs/solana-program-library/pull/2608#discussion_r760681610)